src/oneprometheus/opennebula-exporter/src/opennebula_power_collector.rb

```diff
@@ -0,0 +1,261 @@
+#   gets each process number of CPU ticks per unit of time.
+class OpenNebulaPowerCollector
+
+    HOST_LABELS = [:one_host_id]
+    VM_LABELS = [:one_vm_id]
+    VMPROC_LABELS = [:one_vm_id] + [:proc_pid] + [:proc_name]
+
+    # --------------------------------------------------------------------------
+    # Host power metrics
+    # --------------------------------------------------------------------------
+    #   - opennebula_host_power_consumption_uW
+    # --------------------------------------------------------------------------
+    HOST_POWER_METRICS = {
+        'host_power_consumption_uW' => {
+            :type   => :gauge,
+            :docstr => 'Scaphandre power usage per host uW',
+            :labels => HOST_LABELS
+        }
+    }
+
+    # --------------------------------------------------------------------------
+    # VM power metrics
+    # --------------------------------------------------------------------------
+    #   - opennebula_vm_power_consumption_uW
+    # --------------------------------------------------------------------------
+
+    VM_POWER_METRICS = {
+      'vm_power_consumption_uW' => {
+          :type   => :gauge,
+          :docstr => 'Scaphandre power usage per vm uW',
+          :labels => VM_LABELS
+        }
+    }
+
+    # --------------------------------------------------------------------------
+    # VM proc power metrics
+    # --------------------------------------------------------------------------
+    #   - opennebula_vmproc_power_consumption_uW
+    # --------------------------------------------------------------------------
+    DEFAULT_PROC_INTERVAL = 5
+    DEFAULT_PROC_OVERHEAD = 0.1
+    DEFAULT_PROC_TICKER = '/var/lib/one-context/get_proc_ticks'
+
+    VMPROC_POWER_METRICS = {
+        'vmproc_power_consumption_uW' => {
+            :type   => :gauge,
+            :docstr => 'VM process estimated power usage',
+            :labels => VMPROC_LABELS
+        }
+    }
+
+    VMPROC_PERCENT_METRICS = {
+        'vmproc_cpu_time_percent' => {
+            :type   => :gauge,
+            :docstr => 'VM process CPU time percent',
+            :labels => VMPROC_LABELS
+        }
+    }
+
+    class << self
+        attr_accessor :scaph_data
+    end
+
+    def get_scaph_metrics(host, port, seconds = 2)
+        hostname = String(host['NAME'])
+        Timeout.timeout(seconds) do
+            begin
+                TCPSocket.new(hostname, port).close
+                begin
+                    res = Net::HTTP.get(hostname, '/metrics', port)
+                    self.class.scaph_data[host['ID'].to_i] = res.split("\n")
+                rescue StandardError => e
+                    puts e.message
+                    puts "Error accessing to scaphandre metrics on #{hostname}"
+                end
+            rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH
+                puts "Error on #{hostname}:#{port}, is scaphandre running there?"
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential log injection detected. Ensure all untrusted input is properly sanitized before logging. Use parameterized logging or validate input against an allow list to prevent log injection vulnerabilities. Consider using a dedicated logging library&#39;s built-in sanitization features when available. Learn more - https://cwe.mitre.org/data/definitions/117.html